src/main/java/oshi/hardware/platform/mac/MacSensors.java

```diff
@@ -0,0 +1,305 @@
+            return result;
+        }
+
+        service = IOKit.INSTANCE.IOServiceGetMatchingService(masterPort.getValue(),
+                IOKit.INSTANCE.IOServiceMatching("AppleSMC"));
+        if (service == 0) {
+            LOG.error("Error: no SMC found\n");
+            return result;
+        }
+
+        result = IOKit.INSTANCE.IOServiceOpen(service, SystemB.INSTANCE.mach_task_self(), 0, this.conn);
+        IOKit.INSTANCE.IOObjectRelease(service);
+        if (result != 0) {
+            LOG.error(String.format("Error: IOServiceOpen() = 0x%08x", result));
+            return result;
+        }
+        return 0;
+    }
+
+    private int smcClose() {
+        return IOKit.INSTANCE.IOServiceClose(this.conn.getValue());
+    }
+
+    /*
+     * Get Temperature from SMC
+     */
+    private double smcGetTemperature(String key, int retries) {
+        SMCVal val = new SMCVal();
+        int result = smcReadKey(key, val, retries);
+        if (result == 0) {
+            int temp = 0;
+            if (val.dataSize > 0 && Arrays.equals(val.dataType, IOKit.DATATYPE_SP78)) {
+                temp = (val.bytes[0] * 256 + val.bytes[1]) >> 2;
+            }
+            return temp / 64.0;
+        }
+        // Read failed
+        return 0d;
+    }
+
+    /*
+     * Get number of fans
+     */
+    private int smcGetFanNumber(int retries) {
+        SMCVal val = new SMCVal();
+        int result = smcReadKey(IOKit.SMC_KEY_FAN_NUM, val, retries);
+        if (result == 0) {
+            return strtoul10(val.bytes, val.dataSize);
+        }
+        // Read failed
+        return 0;
+    }
+
+    /*
+     * Get fan speed
+     */
+    private float smcGetFanSpeed(int fanNum, int retries) {
+        SMCVal val = new SMCVal();
+        String key = String.format(IOKit.SMC_KEY_FAN_SPEED, fanNum);
+        int result = smcReadKey(key, val, retries);
+        if (result == 0) {
+            return strtof(val.bytes, val.dataSize, 2);
+        }
+        // Read failed
+        return 0f;
+    }
+
+    /*
+     * Get cached keyInfo if it exists, or generate new
+     */
+    private int smcGetKeyInfo(SMCKeyData inputStructure, SMCKeyData outputStructure) {
+        if (keyInfoCache.containsKey(inputStructure.key)) {
+            outputStructure.keyInfo = keyInfoCache.get(inputStructure.key);
+        } else {
+            int result = 0;
+            inputStructure.data8 = IOKit.SMC_CMD_READ_KEYINFO;
+
+            result = smcCall(IOKit.KERNEL_INDEX_SMC, inputStructure, outputStructure);
+            if (result != 0) {
+                return result;
+            }
+        }
+        return 0;
+    }
+
+    /*
+     * Read a key from SMC
+     */
+    private int smcReadKey(String key, SMCVal val, int retries) {
+        SMCKeyData inputStructure = new SMCKeyData();
+        SMCKeyData outputStructure = new SMCKeyData();
+
+        inputStructure.key = strtoul16(key, 4);
+        int result;
+        do {
+            Util.sleep(8);
```

Comment:
This can't be right :) I would do it in a loop until it succeeds if that is more reliable.